### PR TITLE
Refactor crf code to support constrained decoding

### PR DESCRIPTION
**Pull request by napsternxg**
Fixes tensorflow/addons#607

- We can have a common function for creating `filtered_inputs` used in
`crf_multitag_sequence_score`
- This function can be reused to modify the input to `crf_decode` to
support constrained decoding.

### CODE REVIEW

**Comment by WindQAQ** [comment]
cc @Squadrick

File: tensorflow_addons/text/crf.py [hunk 1]
```diff
@@ -24,6 +24,35 @@
...
+    def _single_seq_fn():
+        filtered_inputs = tf.where(tag_bitmap, inputs,
+                                   tf.fill(tf.shape(inputs), float("-inf")))
+        return filtered_inputs
+
+    def _multi_seq_fn():
+        # Compute the logsumexp of all scores of sequences matching the given tags.
+        filtered_inputs = tf.where(tag_bitmap, inputs,
+                                   tf.fill(tf.shape(inputs), float("-inf")))
+        return filtered_inputs
```
Comment (Squadrick) [inline tensorflow_addons/text/crf.py]: Both `_single_seq_fn` and `_multi_seq_fn` are doing the same thing, you can just do 
```python
def crf_filtered_inputs(inputs, tag_bitmap):
    filtered_inputs = tf.where(tag_bitmap, inputs, tf.fill(tf.shape(inputs), float("-inf")))
    return filtered_inputs
```

Comment (napsternxg) [reply tensorflow_addons/text/crf.py]: Thanks @Squadrick I will try to get back to this next week.

**Comment by napsternxg** [comment]
Hi @Squadrick I have added the tests and made the necessary fixes.

File: tensorflow_addons/text/crf.py [hunk 2]
```diff
@@ -95,22 +116,19 @@ def crf_multitag_sequence_score(inputs, tag_bitmap, sequence_lengths,
 
     # If max_seq_len is 1, we skip the score calculation and simply gather the
     # unary potentials of all active tags.
+    filtered_inputs = crf_filtered_inputs(inputs, tag_bitmap)
```
Comment (Squadrick) [inline tensorflow_addons/text/crf.py]: Where are you getting `tag_bitmap` from?

Comment (napsternxg) [reply tensorflow_addons/text/crf.py]: You need to look at the signature of `crf_multitag_sequence_score` it is passed as an argument.

Comment (napsternxg) [reply tensorflow_addons/text/crf.py]: It is part of the signature of `crf_multitag_sequence_score`

Comment (Squadrick) [reply tensorflow_addons/text/crf.py]: My bad, the code above this line was cut off by GitHub. I thought this was part of `crf_sequence_score`.

Comment (napsternxg) [reply tensorflow_addons/text/crf.py]: Thanks

**Comment by seanpmorgan** [comment]
> @napsternxg Thanks for the update.
> 
> I'd like to wait until #697 is merged before we proceed with this. I'll wrap that up in a day or two.

https://github.com/tensorflow/addons/pull/800 has been merged so I think this is ready for review/merge after fixing merge conflicts.

**Comment by napsternxg** [comment]
@seanpmorgan I have fixed the merge conflicts.

**Comment by napsternxg** [comment]
This PR has become messed up because of the merging with master. I am not aware of how to do this cleanly. I will close this one and reopen a new PR.

**Pull request closed without merging**